src/osd/OSD.cc

```diff
@@ -2160,8 +2154,8 @@ OSD::OSD(CephContext *cct_, ObjectStore *store_,
   objecter_messenger(osdc_messenger),
   monc(mc),
   mgrc(cct_, client_messenger, &mc->monmap),
-  logger(NULL),
-  recoverystate_perf(NULL),
+  logger(create_logger()),
+  recoverystate_perf(create_recoverystate_perf()),
```

Comment:
because `get_osdmap_epoch()` is used for printing the current osdmap epoch in `OSD::create_logger()` and in `OSD::create_recoverystate_perf()`, while osdmap is initialized **after** `logger` and `recoverystate_perf`. valgrind emitted warnings like

```
<error>
  <unique>0x0</unique>
  <tid>1</tid>
  <kind>UninitCondition</kind>
  <what>Conditional jump or move depends on uninitialised value(s)</what>
...
    <frame>
      <ip>0x791D98</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>atomic_load&lt;const OSDMap&gt;</fn>
      <dir>/usr/include/c++/8/bits</dir>
      <file>shared_ptr_atomic.h</file>
      <line>107</line>
    </frame>
    <frame>
      <ip>0x791D98</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>get_osdmap</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3411.g73dcd530b88.el8.x86_64/src/osd</dir>
      <file>OSD.h</file>
      <line>1686</line>
    </frame>
    <frame>
      <ip>0x791D98</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>OSD::get_osdmap_epoch() const</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3411.g73dcd530b88.el8.x86_64/src/osd</dir>
      <file>OSD.h</file>
      <line>1690</line>
    </frame>
    <frame>
      <ip>0x72842B</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>OSD::create_logger()</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3411.g73dcd530b88.el8.x86_64/src/osd</dir>
      <file>OSD.cc</file>
      <line>4048</line>
    </frame>
    <frame>
      <ip>0x781D2E</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>OSD::OSD(ceph::common::CephContext*, ObjectStore*, int, Messenger*, Messenger*, Messenger*, Messenger*, Messenger*, Messenger*, Messenger*, MonClient*, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;, ceph::async::io_context_pool&amp;)</fn>
      <dir>/usr/src/debug/ceph-16.0.0-3411.g73dcd530b88.el8.x86_64/src/osd</dir>
      <file>OSD.cc</file>
      <line>2157</line>
    </frame>
...
```

so i think we have at least three options here

1. do not print debugging messages like `dout(10) << "create_logger" << dendl;` in the beginning of these functions.
1. move declarations of `logger` and `recoverystate_perf` in `OSD` class after that of `_osdmap`
1. keep existing implementation unchanged -- initialize `logger` and `recoverystate_perf` with nullptr. and set them in the body of ctor function.

i am in favor of the first option. what do you think?